File: vaadin-signaturefield/src/main/java/eu/maxschuster/vaadin/signaturefield/internal/SignatureFieldExtension.java
/*
 * Copyright 2015 Max Schuster.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package eu.maxschuster.vaadin.signaturefield.internal;

import com.vaadin.annotations.JavaScript;
import com.vaadin.annotations.StyleSheet;
import com.vaadin.server.AbstractJavaScriptExtension;
import com.vaadin.shared.util.SharedUtil;
import com.vaadin.ui.Component;
import com.vaadin.ui.JavaScriptFunction;
import com.vaadin.util.ReflectTools;
import elemental.json.JsonArray;
import elemental.json.JsonString;
import elemental.json.JsonValue;
import eu.maxschuster.vaadin.signaturefield.SignatureField;
import eu.maxschuster.vaadin.signaturefield.shared.MimeType;
import eu.maxschuster.vaadin.signaturefield.shared.internal.SignatureFieldExtensionState;
import java.lang.reflect.Method;
import eu.maxschuster.vaadin.signaturefield.Color;

/**
 * A javascript extension that extends a {@link SignatureField} with the
 * client-side logic.
 * 
 * <a href="https://github.com/szimek/signature_pad">signature_pad</a> 
 * by Szymon Nowak (<a href="https://github.com/szimek">szimek</a>) is used to
 * capture the signature at the client-side.
 * @author Max Schuster
 * @see <a href="https://github.com/szimek/signature_pad">signature_pad</a>
 */
@JavaScript("vaadin://addons/signaturefield/dist/SignatureFieldExtension.min.js")
@StyleSheet("vaadin://addons/signaturefield/dist/SignatureFieldExtension.css")
public class SignatureFieldExtension extends AbstractJavaScriptExtension {
    
    /**
     * Current signature value
     */
    private String signature;
    
    /**
     * Listener that gets called when the signature changes
     */
    public interface SignatureChangeListener {
        
        public static final Method METHOD = ReflectTools.findMethod(
                SignatureChangeListener.class, "signatureChange",
                SignatureChangeEvent.class);
        
        public void signatureChange(SignatureChangeEvent event);
        
    }
    
    /**
     * A signature change event
     */
    public static class SignatureChangeEvent extends Component.Event {
        
        private final SignatureFieldExtension extension;
        
        private final String signature;

        public SignatureChangeEvent(Component source,
                SignatureFieldExtension extension, String signature) {
            super(source);
            this.extension = extension;
            this.signature = signature;
        }

        /**
         * @return The extension that has fired this event
         */
        public SignatureFieldExtension getExtension() {
            return extension;
        }

        /**
         * @return The new signature value
         */
        public String getSignature() {
            return signature;
        }
        
    }

    public SignatureFieldExtension(SignatureField target) {
        // extend the target
        super(target);
        
        /*
         * Gets called from the client-side when it wants to change the signatue
         * value at the server-side.
         */
        addFunction("fireSignatureChange", new JavaScriptFunction() {

            @Override
            public void call(JsonArray arguments) {
                String signature;
                JsonValue jsonValue = arguments.get(0);
                if (jsonValue instanceof JsonString) {
                    signature = jsonValue.asString();
                } else {
                    signature = null;
                }
                setSignature(signature, true);
                fireSignatureChangeEvent(signature);
            }
        });
    }

    @Override
    protected SignatureFieldExtensionState getState() {
        return (SignatureFieldExtensionState) super.getState();
    }

    @Override
    protected SignatureFieldExtensionState getState(boolean markAsDirty) {
        return (SignatureFieldExtensionState) super.getState(markAsDirty);
    }
    
    @Override
    public void beforeClientResponse(boolean initial) {
        super.beforeClientResponse(initial);
        if (initial) {
            updateSignature();
        }
    }
    
    /**
     * Updates the client-side with the current signature value
     */
    protected void updateSignature() {
        callFunction("setSignature", getSignature());
    }
    
    /**
     * @return Current signature value
     */
    public String getSignature() {
        return signature;
    }
    
    /**
     * Set signature current signature value.
     * @param signature Signature
     * @param repaintIsNotNeeded Repaint is not needed
     */
    public void setSignature(String signature, boolean repaintIsNotNeeded) {
        String oldSignature = this.signature;
        if (!SharedUtil.equals(oldSignature, signature)) {
            this.signature = signature;
            if (!repaintIsNotNeeded) {
                updateSignature();
            }
        }
    }
    
    /**
     * Set signature current signature value.
     * @param signature Signature
     */
    public void setSignature(String signature) {
        setSignature(signature, false);
    }
    
    /**
     * Fires a new {@link SignatureChangeEvent} with the given Signature.
     * @param signature New signature
     */
    public void fireSignatureChangeEvent(String signature) {
        fireEvent(new SignatureChangeEvent((Component) getParent(), this, signature));
    }
    
    /**
     * Adds a {@link SignatureChangeListener}.
     * @param listener Listener to add
     */
    public void addSignatureChangeListener(SignatureChangeListener listener) {
        addListener(SignatureChangeEvent.class, listener, SignatureChangeListener.METHOD);
    }
    
    /**
     * Removes a {@link SignatureChangeListener}.
     * @param listener Listener to remove
     */
    public void removeSignatureChangeListener(SignatureChangeListener listener) {
        removeListener(SignatureChangeEvent.class, listener, SignatureChangeListener.METHOD);
    }
    
    /**
     * Returns true if the extension is immediate.
     * @return Extension is immediate.
     */
    public boolean getImmediate() {
        return getState(false).immediate;
    }
    
    /**
     * Sets the immediate value of this extension.
     * @param immediate New immediate value of this extension.
     */
    public void setImmediate(boolean immediate) {
        getState().immediate = immediate;
    }
    
    /**
     * Returns true if extension is read only.
     * @return Extension is read only.
     */
    public boolean getReadOnly() {
        return getState(false).readOnly;
    }
    
    /**
     * Sets the read only value of this extension.
     * @param readOnly Extension is read only.
     */
    public void setReadOnly(boolean readOnly) {
        getState().readOnly = readOnly;
    }
    
    /**
     * Gets the radius of a single dot.
     *
     * @return Radius of a single dot.
     */
    public Double getDotSize() {
        return getState(false).dotSize;
    }

    /**
     * Sets the radius of a single dot.
     *
     * @param dotSize Radius of a single dot.
     */
    public void setDotSize(Double dotSize) {
        getState().dotSize = dotSize;
    }

    /**
     * Gets the minimum width of a line. Defaults to 0.5.
     *
     * @return Minimum width of a line.
     */
    public double getMinWidth() {
        return getState(false).minWidth;
    }

    /**
     * Sets the minimum width of a line. Defaults to 0.5.
     *
     * @param minWidth Minimum width of a line.
     */
    public void setMinWidth(double minWidth) {
        getState().minWidth = minWidth;
    }

    /**
     * Gets the maximum width of a line.
     *
     * @return Maximum width of a line.
     */
    public double getMaxWidth() {
        return getState(false).maxWidth;
    }

    /**
     * Sets the maximum width of a line.
     *
     * @param maxWidth Maximum width of a line.
     */
    public void setMaxWidth(double maxWidth) {
        getState().maxWidth = maxWidth;
    }

    /**
     * Gets the color used to clear the background. Can be any color format
     * accepted by context.fillStyle. Defaults to "rgba(0,0,0,0)" (transparent
     * black). Use a non-transparent color e.g. "rgb(255,255,255)" (opaque
     * white) if you'd like to save signatures as JPEG images.<br>
     * <br>
     * Some predefined colors can be found in class {@link Color}
     *
     * @return Color used to clear the background.
     */
    public String getBackgroundColor() {
        return getState(false).backgroundColor;
    }

    /**
     * Sets the color used to clear the background. Can be any color format
     * accepted by context.fillStyle. Defaults to "rgba(0,0,0,0)" (transparent
     * black). Use a non-transparent color e.g. "rgb(255,255,255)" (opaque
     * white) if you'd like to save signatures as JPEG images.<br>
     * <br>
     * Some predefined colors can be found in class {@link Color}
     *
     * @param backgroundColor Color used to clear the background.
     */
    public void setBackgroundColor(String backgroundColor) {
        getState().backgroundColor = backgroundColor;
    }

    /**
     * Sets the color used to draw the lines. Can be any color format accepted
     * by context.fillStyle.<br>
     * <br>
     * Some predefined colors can be found in class {@link Color}
     *
     * @return The color used to draw the lines.
     */
    public String getPenColor() {
        return getState(false).penColor;
    }

    /**
     * Sets the color used to draw the lines. Can be any color format accepted
     * by context.fillStyle.<br>
     * <br>
     * Some predefined colors can be found in class {@link Color}
     *
     * @param penColor The color used to draw the lines.
     */
    public void setPenColor(String penColor) {
        getState().penColor = penColor;
    }

    /**
     * Gets the velocity filter weight
     *
     * @return The velocity filter weight
     */
    public double getVelocityFilterWeight() {
        return getState(false).velocityFilterWeight;
    }

    /**
     * Sets the velocity filter weight
     *
     * @param velocityFilterWeight The velocity filter weight
     */
    public void setVelocityFilterWeight(double velocityFilterWeight) {
        getState().velocityFilterWeight = velocityFilterWeight;
    }

    /**
     * Sets the {@link MimeType} of generated images
     *
     * @return The {@link MimeType} of generated images
     */
    public MimeType getMimeType() {
        return MimeType.valueOfMimeType(getState(false).mimeType);
    }

    /**
     * Sets the {@link MimeType} of generated images
     *
     * @param mimeType The {@link MimeType} of generated images
     */
    public void setMimeType(MimeType mimeType) {
        getState().mimeType = mimeType.getMimeType();
    }

    /**
     * Gets the visibility of the clear button
     *
     * @return Should show a clear button in the {@link SignatureFieldExtension}
     */
    public boolean isClearButtonEnabled() {
        return getState(false).clearButtonEnabled;
    }

    /**
     * Sets the visibility of the clear button
     *
     * @param clearButtonEnabled Should show a clear button in the
     * {@link SignatureFieldExtension}
     */
    public void setClearButtonEnabled(boolean clearButtonEnabled) {
        getState().clearButtonEnabled = clearButtonEnabled;
    }
    
}

File: vaadin-signaturefield/src/main/grunt/signaturefield/src/js/SignatureFieldExtension.js
/*
 * Copyright 2015 Max Schuster.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Client-side logic of the SignatureField
 * 
 * @property {Element} parent Parent element
 * @property {Element} canvas Canvas
 * @property {Element} clearButton Button that clears the field
 * @property {string} signature Signature data url
 * @property {string} mimeType MIME-Type that determines the format of the signature image
 * @property {SignaturePad} signaturePad SignaturePad instance
 * @property {boolean} immediate Update the signature on the server-side
 * immediately after each pen stroke
 * @property {boolean} readOnly Field is readonly
 * @author Max Schuster
 * @returns {undefined}
 */
function eu_maxschuster_vaadin_signaturefield_SignatureFieldExtension() {

    "use strict";

    /* jshint validthis:true */
    /* jshint -W087 */ // All 'debugger' statements should be removed

    /**
     * Initializes the extension
     * @returns {undefined}
     */
    this.init = function () {
        var state = this.getState();

        this.parent = this.findAndExtendParent();
        this.canvas = this.createCanvas(this.parent);
        this.clearButton = null;
        this.signature = null;
        this.mimeType = "image/png";
        this.signaturePad = this.createSignaturePad(this.canvas);
        this.immediate = state.immediate;
        this.readOnly = state.readOnly;

        this.addResizeListener(this.parent, this.proxy(this.onResize));
    };

    /**
     * Proxy the given function with a context
     * @param {Function} func Function to proxy
     * @param {Object} [ctx=this] Context to apply to the function. Defaults to
     * "this".
     * @param {Array} [args=undefined] arguments that should be passed to the
     * function. Defaults to "arguments"
     * @returns {Function} The function proxy
     */
    this.proxy = function (func, ctx, args) {
        ctx = ctx || this;
        return function () {
            func.apply(ctx, args || arguments);
        };
    };

    /**
     * Imports the given signature dataURL into the canvas. Called from the
     * server-side to update the signature on the client-side
     * @param {string} newSignature Signture to import
     * @returns {undefined}
     */
    this.setSignature = function (newSignature) {
        var signaturePad = this.signaturePad,
                oldSignature = this.signature;
        if (newSignature === oldSignature) {
            return;
        }
        this.signature = newSignature;
        signaturePad.clear();
        if (newSignature) {
            signaturePad.fromDataURL(newSignature);
        }
    };

    /**
     * Extends and extends the parent component of this extension.
     * @returns {Element} The parent element.
     */
    this.findAndExtendParent = function () {
        var parent = this.getElement(this.getParentId());
        parent.tabIndex = -1;
        this.addEvent(parent, "focus", this.proxy(this.onFocus));
        this.addEvent(parent, "blur", this.proxy(this.onBlur));
        return parent;
    };

    /**
     * Creates the canvas element and attaches it to the given parent element.
     * @param {Element} parent Parent element of the canvas.
     * @returns {unresolved}
     */
    this.createCanvas = function (parent) {
        var canvas = document.createElement("canvas");
        parent.appendChild(canvas);
        return canvas;
    };

    /**
     * Updates the canvas size to match its parent's size, without clearing it.
     * @returns {undefined}
     */
    this.updateCanvasSize = function () {
        var style = this.getComputedStyle(this.parent),
                canvas = this.canvas,
                signature = this.signature,
                signaturePad = this.signaturePad,
                width = parseInt(style.width),
                height = parseInt(style.height);

        canvas.width = width;
        canvas.height = height;

        signaturePad.clear();

        if (signature) {
            signaturePad.fromDataURL(signature);
        }
    };

    /**
     * Creates a signature pad instance with the given canvas element.
     * @param {Element} canvas The canvas to use with the signature pad.
     * @returns {SignaturePad} Signature pad instance.
     */
    this.createSignaturePad = function (canvas) {
        var signaturePad = new SignaturePad(canvas);
        signaturePad.onBegin = this.proxy(this.onStrokeBegin);
        signaturePad.onEnd = this.proxy(this.onStrokeEnd);
        signaturePad.vReadOnly = false;

        // Remove window.devicePixelRatio
        signaturePad.fromDataURL = function (dataUrl) {
            var self = this,
                    image = new Image(),
                    width = this._canvas.width,
                    height = this._canvas.height;

            this._reset();
            image.src = dataUrl;
            image.onload = function () {
                self._ctx.drawImage(image, 0, 0, width, height);
            };
            this._isEmpty = false;
        };

        return signaturePad;
    };

    /**
     * Creates the close button element and adds it to the given parent.
     * @param {Element} parent Parent element to add the close button to.
     * @returns {Element} Close button
     */
    this.createCloseButton = function (parent) {
        var closeButton = document.createElement("button");
        closeButton.setAttribute("class", "signaturefield-clearbutton");
        parent.appendChild(closeButton);
        this.addEvent(closeButton, "click", this.proxy(this.onClearButtonClick));
        return closeButton;
    };

    /**
     * Gets called when the parent element has changed its size.
     * @returns {undefined}
     */
    this.onResize = function () {
        this.updateCanvasSize();
    };

    /**
     * Gets called when a stroke begins.
     * @returns {undefined}
     */
    this.onStrokeBegin = function () {
        this.parent.focus();
    };

    /**
     * Gets called when a stroke ends.
     * @returns {undefined}
     */
    this.onStrokeEnd = function () {
        if (this.immediate) {
            this.updateSignature();
        }
    };

    /**
     * Gets called when the field becomes focused.
     * @returns {undefined}
     */
    this.onFocus = function () {
        this.addClass(this.parent, "v-focus");
    };

    /**
     * Gets called when the field loses focus.
     * @returns {undefined}
     */
    this.onBlur = function () {
        this.updateSignature();
        this.removeClass(this.parent, "v-focus");
    };

    /**
     * Gets called when the clear button is clicked.
     * @returns {undefined}
     */
    this.onClearButtonClick = function () {
        this.clear();
    };
    
    /**
     * Gets called when the shared state of this extension changes. Updates the
     * parts of the extension with the new values from the state.
     * @returns {undefined}
     */
    this.onStateChange = function () {
        var state = this.getState(),
                parent = this.parent,
                signaturePad = this.signaturePad;

        signaturePad.dotSize = state.dotSize;
        signaturePad.minWidth = state.minWidth;
        signaturePad.maxWidth = state.maxWidth;
        signaturePad.backgroundColor = state.backgroundColor;
        signaturePad.penColor = state.penColor;
        signaturePad.velocityFilterWeight = state.velocityFilterWeight;

        this.mimeType = state.mimeType;
        this.immediate = state.immediate;

        var readOnly = state.readOnly,
                vReadOnly = signaturePad.vReadOnly;
        if (readOnly && !vReadOnly) {
            signaturePad.off();
            parent.removeAttribute("tabindex");
        } else if (!readOnly && vReadOnly) {
            signaturePad.on();
            parent.tabIndex = -1;
        }
        signaturePad.vReadOnly = readOnly;

        if ((!readOnly && state.clearButtonEnabled) && !this.clearButton) {
            this.clearButton = this.createCloseButton(parent);
        } else if ((readOnly || !state.clearButtonEnabled) && this.clearButton) {
            var clearButton = this.clearButton;
            parent.removeElement(clearButton);
            this.clearButton = null;
        }
    };

    /**
     * Returns the computed style for the given element
     * @param {Element} el Element to get the styles from
     * @returns {CSSStyleDeclaration}
     */
    this.getComputedStyle = function (el) {
        if (window.getComputedStyle) {
            return window.getComputedStyle(el);
        }
        return el.currentStyle;
    };
    
    /**
     * Gets the current signature as dataURL or null if the signatue pad is
     * empty.
     * @returns {string|null} Signature as dataURL or null.
     */
    this.getCurrentSignature = function () {
        var signaturePad = this.signaturePad;
        return signaturePad.isEmpty() ?
                null : signaturePad.toDataURL(this.mimeType, 1);
    };

    /**
     * Updates the server-side signature value with the given signature.
     * @param {string|null} signature Signature value for the server-side.
     * @returns {undefined}
     */
    this.updateSignature = function (signature) {
        var oldSignature = this.signature,
                newSignature = arguments.length > 0 ?
                signature : this.getCurrentSignature();
        if (newSignature !== oldSignature) {
            this.signature = newSignature;
            this.fireSignatureChange(newSignature);
        }
    };

    /**
     * Clears the signature pad and the signature on the server-side.
     * @returns {undefined}
     */
    this.clear = function () {
        this.signaturePad.clear();
        this.updateSignature(null);
    };

    /**
     * Cross browser add event.
     * Borrowed from https://github.com/samie/Idle
     * 
     * @param {Element} ob
     * @param {string} type
     * @param {Function} fn
     * @returns {Boolean}
     */
    this.addEvent = function (ob, type, fn) {
        if (ob.addEventListener) {
            ob.addEventListener(type, fn, false);
        } else if (ob.attachEvent) {
            ob.attachEvent('on' + type, fn);
        } else {
            type = 'on' + type;
            if (typeof ob[type] === 'function') {
                fn = (function (f1, f2) {
                    return function () {
                        f1.apply(this, arguments);
                        f2.apply(this, arguments);
                    };
                })(ob[type], fn);
            }
            ob[type] = fn;
            return true;
        }
        return false;
    };

    /**
     * Adds a style class to the given element.
     * @param {Element} el Element the class should be added to.
     * @param {string} clazz The class that should be added.
     * @returns {undefined}
     */
    this.addClass = function (el, clazz) {
        var existing = (el.getAttribute("class") || "").split(" ");
        for (var e = 0; e < existing.length; e++) {
            var eClazz = existing[e];
            if (eClazz === clazz) {
                return;
            }
        }
        existing.push(clazz);
        el.setAttribute("class", existing.join(" "));
    };

    /**
     * Removes a style class from the given element.
     * @param {Element} el Element the class should be removed from.
     * @param {string} clazz The class that should be removed.
     * @returns {undefined}
     */
    this.removeClass = function (el, clazz) {
        var existing = (el.getAttribute("class") || "").split(" "),
                length = existing.length;
        for (var e = 0; e < length; e++) {
            var eClazz = existing[e];
            if (eClazz === clazz) {
                existing.splice(e, 1);
                length--;
                e--;
            }
        }
        el.setAttribute("class", existing.join(" "));
    };

    // call the init method
    this.init();

}